### ISSUE TITLE
nnvm doesn't work with OpenCL while tvm does

### ISSUE THREAD

**Issue by kpot** [dmlc/nnvm#172]
Hello guys,

I'm very passionate about using nnvm and mxnet particularly on some of my OpenCL hardware (AMD GPUs). I also have some experience in OpenCL programming and using ML frameworks, so couldn't pass the news about releasing nnvm.
But I cannot make work even simplest nnvm example like this one with OpenCL (the example is based on "get_started.py" tutorial, where the target originally was "cuda").

```
import nnvm.compiler
import nnvm.symbol as sym

x = sym.Variable("x")
y = sym.Variable("y")
z = sym.elemwise_add(x, sym.sqrt(y))
compute_graph = nnvm.graph.create(z)

shape = (4,)
deploy_graph, lib, params = nnvm.compiler.build(
        compute_graph, target="opencl", shape={"x": shape}, dtype="float32")
```

which throws

```
Traceback (most recent call last):
  File "nnvm_test.py", line 11, in <module>
    compute_graph, target="opencl", shape={"x": shape}, dtype="float32")
  File ".../python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/compiler/build_module.py", line 232, in build
    graph = graph.apply("GraphFusePartition").apply("GraphFuseCompile")
  File ".../python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/graph.py", line 234, in apply
    check_call(_LIB.NNGraphApplyPasses(self.handle, npass, cpass, ctypes.byref(ghandle)))
  File ".../python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/_base.py", line 72, in check_call
    raise NNVMError(py_str(_LIB.NNGetLastError()))
nnvm._base.NNVMError: TVMCall CFunc Error:
Traceback (most recent call last):
  File ".../python3.6/site-packages/tvm-0.1.0-py3.6-linux-x86_64.egg/tvm/_ffi/_ctypes/function.py", line 54, in cfun
    rv = local_pyfunc(*pyargs)
  File ".../python3.6/site-packages/nnvm-0.8.0-py3.6.egg/nnvm/top/tensor.py", line 15, in _schedule_injective
    assert target.startswith("llvm")
AssertionError
```

which is strange, because tvm works perfectly fine on the same machine, when I launch this code:

```
import tvm
import numpy as np

n = tvm.var('n')
A = tvm.placeholder((n,), name='A')
B = tvm.placeholder((n,), name='B')
C = tvm.compute(A.shape, lambda i: A[i] + B[i], name='C')
print(type(C))

s = tvm.create_schedule(C.op)
bx, tx = s[C].split(C.op.axis[0], factor=64)

s[C].bind(bx, tvm.thread_axis("blockIdx.x"))
s[C].bind(tx, tvm.thread_axis("threadIdx.x"))

fadd_cl = tvm.build(s, [A, B, C], "opencl", name="myadd")

ctx = tvm.cl(0)
n = 1024

a = tvm.nd.array(np.random.uniform(size=n).astype(A.dtype), ctx)
b = tvm.nd.array(np.random.uniform(size=n).astype(B.dtype), ctx)
c = tvm.nd.array(np.zeros(n, dtype=C.dtype), ctx)
fadd_cl(a, b, c)

np.testing.assert_allclose(c.asnumpy(), a.asnumpy() + b.asnumpy())
```

It works correctly and prints out

```
[17:59:31] src/runtime/opencl/opencl_device_api.cc:195: Initialize OpenCL platform 'AMD Accelerated Parallel Processing'
[17:59:31] src/runtime/opencl/opencl_device_api.cc:215: opencl(0)='Tahiti' cl_device_id=0x1eb9470
[17:59:31] src/runtime/opencl/opencl_device_api.cc:215: opencl(1)='Tahiti' cl_device_id=0x21914b0
```

I can even print the kernel's code. So it's definitely not hardware/OpenCL/driver/tvm issue.
I suspect that the problem has something to do with nnvm compiler not knowing for which OpenCL device it should build the kernel. Could you please send me in right direction about what to read or try?

**Comment by tqchen** [comment]
This should be fixed by #173

**Comment by kpot** [comment]
Yes, `get_started.py` tutorial is working now, after replacing target "cuda" with "opencl" and `tvm.gpu(...)` with `tvm.cl(...)`.
Thank you!
